### ISSUE TITLE
Please test 4.0.0b0 Release

### ISSUE THREAD

**Issue by johnthagen** [jazzband/django-dbbackup#415]
A new release `4.0.0b0` has been released onto PyPI: https://pypi.org/project/django-dbbackup/4.0.0b0/

Because it has been over a  year since the last release, the maintainers are asking users to test out this beta release. The main changes are that Django 4.0 is supported and the minimum supported Django is now 2.2, the oldest Django version still supported by the Django Team.

Please test this by running:

```shell
pip install django-dbbackup==4.0.0b0
```

**Comment by johnthagen** [comment]
Pinging `django-dbbackup` users who have noted issues that this release should address:

@pulse-mind @Ajaysainisd @KessoumML @jerinpetergeorge @awtimmering @banagale @Archmonger

**Comment by Archmonger** [comment]
I have been testing `django-dbbackup@9d1909c30a3271c8c9c8450add30d6e0b996e145` for over a week and am happy to report no issues. I will switch to the tagged release `4.0.0b0` but since the changes are minor I don't expect any issues. It would be relatively safe to proceed with a 4.0 release.

**Comment by pulse-mind** [comment]
Hi, I tested on a project. I did a backup before upgrade after after upgrade. I've got a file a little bit bigger but the extension was psql.bin.gz. When I edit it with a txt editor it looks like a pgdump but before I got a psql file.

**Comment by johnthagen** [comment]
I tested `4.0.0b0` on a production application (targeting Postgres 13) and backup and restore worked as expected.

**Comment by johnthagen** [comment]
@Ajaysainisd @KessoumML @jerinpetergeorge @awtimmering @banagale @zwiebelslayer Have any of you been able to test the `4.0.0b0` release? I was hoping to get a bit more feedback before pushing a non-beta release.

@pulse-mind Could you help dissect which commit/PR made the change you are referring to? I didn't see anything super obvious related to compression in the changelog.

**Comment by Archmonger** [comment]
I've gone ahead and confirmed that this package is critically broken.

Although a backup is successfully made via the `dbbackup` command, it cannot be restored via `dbrestore`. This does not appear to be related to compression. 

The console logs will get swarmed with errors which fall into one of these three categories:

1) index already exists
2) syntax error
3) unrecognized token


```cmd
C:\Users\Repositories\Conreq\.venv\lib\site-packages\dbbackup\db\sqlite.py:77: UserWarning: Error in db restore: index silk_profile_queries_profile_id_sqlquery_id_b2403d9b_uniq already exists

C:\Users\Repositories\Conreq\.venv\lib\site-packages\dbbackup\db\sqlite.py:77: UserWarning: Error in db restore: near "579": syntax error
  warnings.warn("Error in db restore: {}".format(err))

C:\Users\Repositories\Conreq\.venv\lib\site-packages\dbbackup\db\sqlite.py:77: UserWarning: Error in db restore: unrecognized token: "'         4802 function calls (4786 primitive calls) in 0.009 seconds
"
```

**Comment by pulse-mind** [comment]
Hi,

Let me explain again because it was not clear enough:
**Test A (on linux on 26 dec 2021)**
* django-dbbackup 3.3.0 : I get a file not compressed (even if I ask compressed) that is a psql file (regular file that contains the requests)
* django-dbbackup 4.0.0b0 : I get a file not compressed (even if I ask compressed) that is a dumpfile
Note : psql (PostgreSQL) 11.14 (Raspbian 11.14-0+deb10u1)

**Test B (on windows on 8 Jan 2022)**
* django-dbbackup 3.3.0 : I get a file compressed (I ask compressed) that is a psql file (regular file that contains the requests)
* django-dbbackup 4.0.0b0 : I get a file compressed (I ask compressed) that is a dumpfile
Note : PostgreSQL 11

So I do not know why but on linux the compress does not work, may be I did not installed all the dependencies!
This commit made the change : https://github.com/jazzband/django-dbbackup/commit/b8d4e2b73b15bc2c2a84f0d58837554e1f656f60 "Make the binary connector the default". 

This is fine for me there is no problem using binary connector by default. 
I do not understand why it does not compress on linux, I would expect a better job on linux compare to windows, usually it is better or easier on linux :D.

**Comment by johnthagen** [comment]
@Archmonger I wonder if your SQLite issue is related to #254 / #258 / #383 ?

Also, could you confirm if  this issue also exists for the 3.3.0 release? e.g. Is this a regression from 3.3.0 -> 4.0.0b0, or is this broken on the latest stable release too?

**Comment by johnthagen** [comment]
@Archmonger I made a very disturbing discovery this morning. I noticed that the GitHub Actions aren't actually running the unit tests for this project 🤦‍♂️. It looks like the original transition to GitHub Actions #368 made a critical typo (`py36` vs `py3.6`) and as a result no environments have been running.

I opened #419 to fix this so that we can at least see the failing tests now. I'll need @jezdez to grant me permission on this repo again so we can force merge it (so that at least we can begin the work of attempting to fix things).

**Comment by johnthagen** [comment]
@Archmonger I think taking a step back, let's get #419 merged and then I think we should consider trying to get the functional tests running in CI as well (#369).

I did confirm that the unit tests do not run for me under Windows, so there is likely at a minimum a Windows/SQLite issue.

**Comment by Archmonger** [comment]
On the note of Windows compatibility, I think we need to have a tag up about the project direction.

Based on the project description it should be using traditional dump and restore mechanisms, but that description may no longer hold water as times have changed.

Additionally, dumpdata supports integrated compression in multiple different formats, so that's another thing for us to look towards.

Shoot me an email at archiethemonger@gmail.com so we can discuss.

**Comment by Archmonger** [comment]
In terms of porting, if we utilize the new Django internals for dumpdata, then we should expect zero backwards compatibility with the older backup files.

If we do take such an extreme approach, the effort of porting vs a new package is fairly insignificant. I think the debate comes down to, do we continue development to support this fairly antiquated backup method, or do we break compatibility in exchange for using Django's new backup internals.

I'm going to create a separate issue ticket specifically to discuss this project's future direction. It will outline what options we can possibly take moving forward.

**Comment by Archmonger** [comment]
@johnthagen Issue has been created for everyone to discuss upon. I think we should keep it open for a month or two to give people time to voice their thoughts.

**Comment by Archmonger** [comment]
@johnthagen do you feel comfortable with me pushing a 4.0 release?

**Comment by johnthagen** [comment]
@Archmonger `4.0.0b0` has been working for *me* in production, but my only concerns are the two comments in this thread:

- https://github.com/jazzband/django-dbbackup/issues/415#issuecomment-1007987348
- https://github.com/jazzband/django-dbbackup/issues/415#issuecomment-1007988364

Given that there aren't any PRs to fix these issues and the beta fixes other issues (such as Django 4 support) I'm inclined to release it. Users can always pin to `3.3.0` if they need to.

Do we have access to ReadTheDocs yet so that we can update the documentation?

**Comment by Archmonger** [comment]
We do not have RTD access yet. I've reached out a couple times and still no response.

**Comment by johnthagen** [comment]
I'd say feel free to go ahead and release 4.0.0 whenever you're ready. We bumped the major version to signal that this is a big change.

**Comment by liorp** [comment]
Waiting for it impatiently 😊 Any update on the release schedule?

**Comment by johnthagen** [comment]
@ZuluPro Could you please grant @Archmonger and myself permissions to the [ReadTheDocs](https://django-dbbackup.readthedocs.io/en/master/) site for this project? We should update the docs when doing the 4.0.0 release.

**Comment by Archmonger** [comment]
I've tried various methods of contact with no success. In the event that we don't hear back from the original maintainers, I propose we rename this repo to django-dbbackup2 upon completing our backend rework.

**Comment by adi-** [comment]
Main functionality is working for me. However, there are some serious problems with logging. I have in my setting:

```
'loggers': {
        'dbbackup': {
            'level': 'WARNING',
            'handlers': ['console', 'file', 'mail_admins'],
            'propagate': False,
        },
        'dbbackup.command': {
            'level': 'WARNING',
            'handlers': ['console', 'file', 'mail_admins'],
            'propagate': False,
        },
```

and I still see INFO logs. 

Changing above to:

```
'loggers': {
        'dbbackup': {
            'level': 'INFO',
            'handlers': ['console', 'file', 'mail_admins'],
            'propagate': False,
        },
```

doesn't log **anything**

**Comment by wilinger** [comment]
Getting SuspiciousFileOperation error using Dropbox as storage from inside a Docker container even thought compressed file uploads successfully and I am able to decompress.

Django 4.0.3
Python 3.10.4
django-dbbackup 4.0.0b0

settings.py
```
DBBACKUP_STORAGE = "storages.backends.dropbox.DropBoxStorage"
DBBACKUP_STORAGE_OPTIONS = {
    "oauth2_access_token": env("DBBACKUP_TOKEN"),
}
```
```
django@b669c732255f:~$ python manage.py dbbackup -z
INFO Backing Up Database: mydb
INFO Writing file to default-b669c732255f-2022-04-04-024226.psql.bin.gz
INFO Request to files/get_metadata                                                                                                                         DEBUG Starting new HTTPS connection (1): api.dropboxapi.com:443
DEBUG https://api.dropboxapi.com:443 "POST /2/files/get_metadata HTTP/1.1" 409 94
INFO Request to files/upload_session/start
DEBUG Starting new HTTPS connection (1): content.dropboxapi.com:443                                                                                        DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/start HTTP/1.1" 200 85
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/append_v2                                                                                                             DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/append_v2 HTTP/1.1" 200 4
INFO Request to files/upload_session/finish
DEBUG https://content.dropboxapi.com:443 "POST /2/files/upload_session/finish HTTP/1.1" 200 None
ERROR SuspiciousFileOperation: Detected path traversal attempt in '/default-b669c732255f-2022-04-04-024226.psql.bin.gz'
  File "/usr/local/lib/python3.10/site-packages/dbbackup/utils.py", line 114, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/dbbackup.py", line 66, in handle
    self._save_new_backup(database)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/dbbackup.py", line 93, in _save_new_backup
    self.write_to_storage(outputfile, filename)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/_base.py", line 82, in write_to_storage
    self.storage.write_file(file, path)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/storage.py", line 82, in write_file
     self.storage.save(name=filename, content=filehandle)
  File "/usr/local/lib/python3.10/site-packages/django/core/files/storage.py", line 59, in save
    validate_file_name(name, allow_relative_path=True)
  File "/usr/local/lib/python3.10/site-packages/django/core/files/utils.py", line 18, in validate_file_name
    raise SuspiciousFileOperation(

Traceback (most recent call last):
  File "/app/manage.py", line 22, in <module>
    main()
  File "/app/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.10/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.10/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.10/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.10/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/utils.py", line 114, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/dbbackup.py", line 66, in handle
    self._save_new_backup(database)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/dbbackup.py", line 93, in _save_new_backup
    self.write_to_storage(outputfile, filename)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/management/commands/_base.py", line 82, in write_to_storage
    self.storage.write_file(file, path)
  File "/usr/local/lib/python3.10/site-packages/dbbackup/storage.py", line 82, in write_file
    self.storage.save(name=filename, content=filehandle)
  File "/usr/local/lib/python3.10/site-packages/django/core/files/storage.py", line 59, in save
    validate_file_name(name, allow_relative_path=True)
  File "/usr/local/lib/python3.10/site-packages/django/core/files/utils.py", line 18, in validate_file_name
    raise SuspiciousFileOperation(
django.core.exceptions.SuspiciousFileOperation: Detected path traversal attempt in '/default-b669c732255f-2022-04-04-024226.psql.bin.gz'
```

**Comment by Archmonger** [comment]
Yeah... That's another downside of this project using it's own backup logic. To my understanding, the backups are funneled through `FileField`, which have `SuspiciousFileOperation` protection against writing to anything besides the media directory.

This problem shouldn't exist when we redesign things. See:
- https://github.com/jazzband/django-dbbackup/issues/423

**Comment by Archmonger** [comment]
I've reached out to RTD and am awaiting their response for ownership transfer.

**Comment by banagale** [comment]
If you can demonstrate control of Pypi publishing and to this repository, RTD should come around on this issue in a responsive manner. Please post if you'd me to also reach out to them.

**Comment by Archmonger** [comment]
I've obtained ownership of the RTD recently. PyPI ownership has also been obtained a while ago.

See this issue for where we are at
- https://github.com/jazzband/django-dbbackup/issues/417

**Comment by banagale** [comment]
Ah! This is great progress. Thank you.

**Comment by pulse-mind** [comment]
Hello, 
What is the plan about the 4.0.0 release using pypi ?

**Comment by banagale** [comment]
Is there a reasonable amount of time before this package should be considered spun back out of jazzband? 

For example, if @johnthagen is interested, is there a process for nominating and liberating a package?

I recognize how challenging maintenance is, and want to give JB time to sort out internal issues. 

However, if this release isn’t unblocked by say August, is it fair to seriously start whatever process is necessary to move this package ahead?

**Comment by Archmonger** [comment]
Yes - A transfer out of Jazzband is very much up for consideration.

However, somewhat ironically, a transfer out of jazzband would be blocked the same issue:
- https://github.com/jazzband/help/issues/196

I should be able to awkwardly circumvent Jazzband for PyPI releases. But, we definitely should still reconsider this package's organization.

**Comment by Archmonger** [comment]
4.0.0 has been manually released, circumventing formal Jazzband procedures.
